core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -495,19 +495,21 @@ class TransactionStateManager(brokerId: Int,
             memRecords.batches.forEach { batch =>
               for (record <- batch.asScala) {
                 require(record.hasKey, "Transaction state log's key should not be null")
-                try {
-                  val transactionalId = TransactionLog.readTxnRecordKey(record.key).asInstanceOf[String]
+                val transactionalId = try Some(TransactionLog.readTxnRecordKey(record.key))
+                catch {
+                  case e: IllegalStateException =>
+                    warn(s"Unknown message key version while loading transaction state from $topicPartition. " +
+                      s"Ignoring it. It could be a left over from an aborted upgrade. Error: ${e.getMessage}")
```

Comment:
```scala
                    warn(s"Unknown message key version while loading transaction state from $topicPartition. " +
                      s"Ignoring it. It could be a left over from an aborted upgrade", e)
```